backend/ranking.js

```diff
@@ -0,0 +1,158 @@
+import { Stock } from "./models/stock.js";
+import { Follow } from "./models/follow.js";
+import { Like } from "./models/like.js";
+import { Dislike } from "./models/dislike.js";
+import { compareStocksByPoints } from "./utils.js";
+
+
+const FOLLOW_POINTS = 25;
+const LIKE_POINTS = 20;
+const DISLIKE_POINTS = -20;
+
+
+export async function RankingOptimized(user) {
+    try {
+        const [stocksInDatabase, follows, likes, dislikes] = await Promise.all([
+            Stock.findAll(),
+            Follow.findAll({
+                where: { UserId: user.id },
+                include: [{ model: Stock }]
+            }),
+            Like.findAll({
+                where: { UserId: user.id },
+                include: [{ model: Stock }]
+            }),
+            Dislike.findAll({
+                where: { UserId: user.id },
+                include: [{ model: Stock }]
+            })
+        ]);
+
+        if (stocksInDatabase.length < 10) {
+            return {
+                status: 422,
+                error: "You need at least 10 stocks in the database to access ranking."
+            };
+        }
+
+        const followsMap = {};
+
+        follows.forEach((follow) => {
+            followsMap[follow.Stock.ticker] = true;
+        });
+
+        const likesMap = {};
+
+        likes.forEach((like) => {
+            likesMap[like.Stock.ticker] = true;
+        });
+
+        const dislikesMap = {};
+
+        dislikes.forEach((dislike) => {
+            dislikesMap[dislike.Stock.ticker] = true;
+        });
+
+        const stocksRanking = stocksInDatabase.map((currentStock) => {
+            let points = 0;
+
+            if (followsMap[currentStock.ticker]) {
+                points += FOLLOW_POINTS;
+            }
+
+            if (likesMap[currentStock.ticker]) {
+                points += LIKE_POINTS;
+            }
+
+            if (dislikesMap[currentStock.ticker]) {
+                points += DISLIKE_POINTS;
+            }
+
+            const { ticker, name, logo } = currentStock.dataValues;
+            const stockWithPoints = { ticker, name, logo, points };
+
+            return stockWithPoints;
+        });
+
+        stocksRanking.sort(compareStocksByPoints);
+        const topTenStocks = stocksRanking.slice(0, 10);
+
+        return { status: 200, data: { stocksRanking: topTenStocks } };
+    } catch (error) {
+        return { status: 500, error };
+    }
+}
+
```

Comment:
This is a big improvement in performance from previous implementation, great job! 

Not a priority, but something interesting to think about is how you can optimize something even further -- Here we are re-evaluating the points for each stock in the DB when we come to the stocks ranking page. Consider this scenario: We have a really large DB of stocks, but only 1 or 2 stocks have been updated (like/dislike/follow). 

In this case, we don't really need to iterate through the entire DB's stocks each time. If the DB stores the points for each stock, it can be updated real-time (ex. when someone likes the stock, increment the points by `LIKE_POINTS` for that particular stock). This makes it easier on the DB when it comes to displaying the full ranking later.